scheduler/system_util_test.go

```diff
@@ -15,13 +15,31 @@ import (
 	"github.com/hashicorp/nomad/nomad/structs"
 )
 
+// diffResultCount is a test helper struct that makes it easier to specify an
+// expected diff
+type diffResultCount struct {
+	place, update, migrate, stop, ignore, lost, disconnecting, reconnecting int
+}
+
+// assertDiffCount is a test helper that compares against a diffResult
+func assertDiffCount(t *testing.T, expected diffResultCount, diff *diffResult) {
+	t.Helper()
+	test.Len(t, expected.update, diff.update, test.Sprintf("expected update"))
+	test.Len(t, expected.ignore, diff.ignore, test.Sprintf("expected ignore"))
+	test.Len(t, expected.stop, diff.stop, test.Sprintf("expected stop"))
+	test.Len(t, expected.migrate, diff.migrate, test.Sprintf("expected migrate"))
+	test.Len(t, expected.lost, diff.lost, test.Sprintf("expected lost"))
+	test.Len(t, expected.place, diff.place, test.Sprintf("expected place"))
+}
```

Comment:
> These are super nice, and I like that they are very simple to use, but is it a concern if we happen to miscount things and, for example, place the same allocation in two categories in a way that the count matches but the actual allocs do not?
>
> I see some assertions being done after `assertDiffCount` to check for this in some cases, but maybe we could do this automatically in the helper here if it took a set of alloc IDs?

That's what I did in my first attempt :grinning:  The problem is that if you have placements, the alloc IDs aren't known to the caller (the test) because they don't exist yet. It kind of ends up being a mess of special cases. I think it'd be worth taking another crack at, but maybe not in this PR?

> Another side note, I think these are useful for non-system schedulers as well, so maybe it would be better to put them in `util_test.go`?

That would totally make sense, except the reconciler tests (used by non-system schedulers) has its own return object and diff test helper. I'd like to get them all using the same diff object but that may need to wait until we do system deployments.